test/sslapitest.c

```diff
@@ -6684,7 +6684,77 @@ static int test_ticket_callbacks(int tst)
  *                  sends CertificateRequest, client reads and ignores it
  * Test 6: TLSv1.3, server continues to read/write after client shutdown, client
  *                  doesn't read it
+ * Test 7: TLSv1.3, client does not shutdown properly,
+ *                  server does not set SSL_OP_IGNORE_UNEXPECTED_EOF,
+ *                  server should get SSL_ERROR_SSL
+ * Test 8: TLSv1.3, client does not shutdown properly,
+ *                  server sets SSL_OP_IGNORE_UNEXPECTED_EOF,
+ *                  server should get SSL_ERROR_ZERO_RETURN
  */
+
+static int test_incorrect_shutdown(int tst)
+{
+    SSL_CTX *cctx = NULL, *sctx = NULL;
+    SSL *clientssl = NULL, *serverssl = NULL;
+    int testresult = 0;
+    char msg[] = "A test message";
+    char buf[80];
+    size_t written, readbytes;
+    SSL_SESSION *sess;
+    BIO *c2s;
+
+    if (!TEST_true(create_ssl_ctx_pair(libctx, TLS_server_method(),
+                                       TLS_client_method(),
+                                       TLS1_VERSION,
+                                       (tst <= 1) ? TLS1_2_VERSION
+                                                  : TLS1_3_VERSION,
+                                       &sctx, &cctx, cert, privkey)))
+        goto end;
+
+    SSL_CTX_set_quiet_shutdown(cctx, 1);
+
+    if (tst == 8)
+        SSL_CTX_set_options(sctx, SSL_OP_IGNORE_UNEXPECTED_EOF);
+
+    if (!TEST_true(create_ssl_objects(sctx, cctx, &serverssl, &clientssl,
+                                            NULL, NULL)))
+        goto end;
+
+    if (!TEST_true(create_ssl_connection(serverssl, clientssl,
+                                              SSL_ERROR_NONE))
+            || !TEST_ptr_ne(sess = SSL_get_session(clientssl), NULL)
+            || !TEST_true(SSL_SESSION_is_resumable(sess)))
+        goto end;
+
+    c2s = SSL_get_rbio(serverssl);
+    BIO_set_mem_eof_return(c2s, 0);
+
+    if (!TEST_true(written = SSL_write(clientssl, msg, sizeof(msg))))
+        goto end;
+    if (!TEST_true(readbytes = SSL_read(serverssl, buf, sizeof(buf))))
+        goto end;
```

Comment:
I'm not sure what you want to do here, but you seem to set written and readbytes, then don't do anything with it.

Reply:
Done. Thanks!